### PR TITLE
tools/edbg: add and use wrapper shell script

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

As requested in #15970, this PR splits out the changes related to edbg in its own PR.

While working on #15970, I noticed that support for edbg was not consistent with the other flashers, using a custom command with a dual call to edbg.
Similar to openocd/jlink/pyocd, this PR adds an edbg.sh script that wrap the flash and reset calls to edbg. As a result, edbg.inc.mk needs some refactoring as well.

The good thing is that there's no need to define a custom `flash-recipe = $(edbg-flash-recipe)`.

### Testing procedure

Flashing and resetting with edbg still works. Also verify that flashing using DEBUG_ADAPTER_ID still works.

Results on samr21-xpro:

<details><summary>flash</summary>

```
$ BUILD_IN_DOCKER=1 make BOARD=samr21-xpro -C examples/hello-world flash --no-print-directory 
Launching build container using image "riot/riotbuild:latest".
docker run --rm --tty --user $(id -u) -v '/usr/share/zoneinfo/Europe/Paris:/etc/localtime:ro' -v '/work/riot/RIOT:/data/riotbuild/riotbase:delegated' -e 'RIOTBASE=/data/riotbuild/riotbase' -e 'CCACHE_BASEDIR=/data/riotbuild/riotbase' -e 'BUILD_DIR=/data/riotbuild/riotbase/build' -e 'RIOTPROJECT=/data/riotbuild/riotbase' -e 'RIOTCPU=/data/riotbuild/riotbase/cpu' -e 'RIOTBOARD=/data/riotbuild/riotbase/boards' -e 'RIOTMAKE=/data/riotbuild/riotbase/makefiles'      -e 'BOARD=samr21-xpro'  -w '/data/riotbuild/riotbase/examples/hello-world/' 'riot/riotbuild:latest' make 'BOARD=samr21-xpro'    
Building application "hello-world" for "samr21-xpro" with MCU "samd21".

"make" -C /data/riotbuild/riotbase/boards/samr21-xpro
"make" -C /data/riotbuild/riotbase/core
"make" -C /data/riotbuild/riotbase/cpu/samd21
"make" -C /data/riotbuild/riotbase/cpu/cortexm_common
"make" -C /data/riotbuild/riotbase/cpu/cortexm_common/periph
"make" -C /data/riotbuild/riotbase/cpu/sam0_common
"make" -C /data/riotbuild/riotbase/cpu/sam0_common/periph
"make" -C /data/riotbuild/riotbase/cpu/samd21/periph
"make" -C /data/riotbuild/riotbase/cpu/samd21/vectors
"make" -C /data/riotbuild/riotbase/drivers
"make" -C /data/riotbuild/riotbase/drivers/periph_common
"make" -C /data/riotbuild/riotbase/sys
"make" -C /data/riotbuild/riotbase/sys/auto_init
"make" -C /data/riotbuild/riotbase/sys/malloc_thread_safe
"make" -C /data/riotbuild/riotbase/sys/newlib_syscalls_default
"make" -C /data/riotbuild/riotbase/sys/pm_layered
"make" -C /data/riotbuild/riotbase/sys/stdio_uart
   text	   data	    bss	    dec	    hex	filename
   9004	    112	   2304	  11420	   2c9c	/data/riotbuild/riotbase/examples/hello-world/bin/samr21-xpro/hello-world.elf
/work/riot/RIOT/dist/tools/edbg/edbg.sh flash /work/riot/RIOT/examples/hello-world/bin/samr21-xpro/hello-world.bin
### Flashing Target ###
Debugger: ATMEL EDBG CMSIS-DAP ATML2127031800009840 01.1A.00FB (S)
Clock frequency: 16.0 MHz
Target: SAM R21G18 (Rev C)
Verification....................................... done.
Done flashing
```

</details>

<details><summary>reset</summary>

```
$ BUILD_IN_DOCKER=1 make BOARD=samr21-xpro -C examples/hello-world reset --no-print-directory 
/work/riot/RIOT/dist/tools/edbg/edbg.sh reset
### Resetting Target ###
```

</details>

<details><summary>flash with DEBUG_ADAPTER_ID</summary>

```
$ DEBUG_ADAPTER_ID=ATML2127031800009840 make BOARD=samr21-xpro -C examples/hello-world flash-only --no-print-directory 
/work/riot/RIOT/dist/tools/edbg/edbg.sh flash /work/riot/RIOT/examples/hello-world/bin/samr21-xpro/hello-world.bin
### Flashing Target ###
Debugger: ATMEL EDBG CMSIS-DAP ATML2127031800009840 01.1A.00FB (S)
Clock frequency: 16.0 MHz
Target: SAM R21G18 (Rev C)
Verification....................................... done.
Done flashing
```

</details>

<details><summary>flash riotboot</summary>

```
$ make BOARD=samr21-xpro -C tests/riotboot riotboot/flash term --no-print-directory 
compiling /work/riot/RIOT/dist/tools/riotboot_gen_hdr/bin/genhdr...
make: Nothing to be done for 'all'.
"make" -C /work/riot/RIOT/boards/samr21-xpro
"make" -C /work/riot/RIOT/core
"make" -C /work/riot/RIOT/cpu/samd21
"make" -C /work/riot/RIOT/cpu/cortexm_common
"make" -C /work/riot/RIOT/cpu/cortexm_common/periph
"make" -C /work/riot/RIOT/cpu/sam0_common
"make" -C /work/riot/RIOT/cpu/sam0_common/periph
"make" -C /work/riot/RIOT/cpu/samd21/periph
"make" -C /work/riot/RIOT/cpu/samd21/vectors
"make" -C /work/riot/RIOT/drivers
"make" -C /work/riot/RIOT/drivers/periph_common
"make" -C /work/riot/RIOT/sys
"make" -C /work/riot/RIOT/sys/auto_init
"make" -C /work/riot/RIOT/sys/checksum
"make" -C /work/riot/RIOT/sys/isrpipe
"make" -C /work/riot/RIOT/sys/malloc_thread_safe
"make" -C /work/riot/RIOT/sys/newlib_syscalls_default
"make" -C /work/riot/RIOT/sys/pm_layered
"make" -C /work/riot/RIOT/sys/riotboot
"make" -C /work/riot/RIOT/sys/shell
"make" -C /work/riot/RIOT/sys/shell/commands
"make" -C /work/riot/RIOT/sys/stdio_uart
"make" -C /work/riot/RIOT/sys/test_utils/interactive_sync
"make" -C /work/riot/RIOT/sys/tsrb
creating /work/riot/RIOT/tests/riotboot/bin/samr21-xpro/tests_riotboot-slot0.1613728886.riot.bin...
/work/riot/RIOT/dist/tools/edbg/edbg.sh flash /work/riot/RIOT/tests/riotboot/bin/samr21-xpro/tests_riotboot-slot0.1613728886.riot.bin
### Flashing Target ###
Debugger: ATMEL EDBG CMSIS-DAP ATML2127031800009840 01.1A.00FB (S)
Clock frequency: 16.0 MHz
Target: SAM R21G18 (Rev C)
Verification...
at address 0x1004 expected 0x76, read 0x49
Error: verification failed
Debugger: ATMEL EDBG CMSIS-DAP ATML2127031800009840 01.1A.00FB (S)
Clock frequency: 16.0 MHz
Target: SAM R21G18 (Rev C)
Programming............................................................ done.
Verification............................................................ done.
Done flashing
Building application "riotboot" for "samr21-xpro" with MCU "samd21".

"make" -C /work/riot/RIOT/boards/samr21-xpro
"make" -C /work/riot/RIOT/core
"make" -C /work/riot/RIOT/cpu/samd21
"make" -C /work/riot/RIOT/cpu/cortexm_common
"make" -C /work/riot/RIOT/cpu/cortexm_common/periph
"make" -C /work/riot/RIOT/cpu/sam0_common
"make" -C /work/riot/RIOT/cpu/sam0_common/periph
"make" -C /work/riot/RIOT/cpu/samd21/periph
"make" -C /work/riot/RIOT/cpu/samd21/vectors
"make" -C /work/riot/RIOT/drivers
"make" -C /work/riot/RIOT/drivers/periph_common
"make" -C /work/riot/RIOT/sys
"make" -C /work/riot/RIOT/sys/checksum
"make" -C /work/riot/RIOT/sys/malloc_thread_safe
"make" -C /work/riot/RIOT/sys/newlib_syscalls_default
"make" -C /work/riot/RIOT/sys/riotboot
"make" -C /work/riot/RIOT/sys/stdio_null
   text	   data	    bss	    dec	    hex	filename
   1644	      0	    592	   2236	    8bc	/work/riot/RIOT/bootloaders/riotboot/bin/samr21-xpro/riotboot.elf
/work/riot/RIOT/dist/tools/edbg/edbg.sh flash /work/riot/RIOT/bootloaders/riotboot/bin/samr21-xpro/riotboot.bin
### Flashing Target ###
Debugger: ATMEL EDBG CMSIS-DAP ATML2127031800009840 01.1A.00FB (S)
Clock frequency: 16.0 MHz
Target: SAM R21G18 (Rev C)
Verification...
at address 0x4 expected 0x81, read 0xb1
Error: verification failed
Debugger: ATMEL EDBG CMSIS-DAP ATML2127031800009840 01.1A.00FB (S)
Clock frequency: 16.0 MHz
Target: SAM R21G18 (Rev C)
Programming.......... done.
Verification.......... done.
Done flashing
/work/riot/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200"  
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2021-02-19 11:01:50,779 # Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
help
2021-02-19 11:01:54,786 # help
2021-02-19 11:01:54,789 # Command              Description
2021-02-19 11:01:54,792 # ---------------------------------------
2021-02-19 11:01:54,797 # curslotnr            Print current slot number
2021-02-19 11:01:54,801 # curslothdr           Print current slot header
2021-02-19 11:01:54,806 # getslotaddr          Print address of requested slot
2021-02-19 11:01:54,810 # dumpaddrs            Prints all slot data in header
2021-02-19 11:01:54,814 # reboot               Reboot the node
2021-02-19 11:01:54,818 # version              Prints current RIOT_VERSION
2021-02-19 11:01:54,823 # pm                   interact with layered PM subsystem
> 2021-02-19 11:01:55,732 # Exiting Pyterm
```

</details>

### Issues/PRs references

split-out from #15970

### CODE REVIEW

**Comment by fjmolinas** [comment]
@aabadie can you address shellcheck comments?

**Comment by fjmolinas** [comment]
> @aabadie can you address shellcheck comments?

Otherwise this looks good!

**Comment by aabadie** [comment]
>  can you address shellcheck comments?

Sorry, I forgot to cherry-pick a fixup commit for from #15970

File: makefiles/tools/edbg.inc.mk [hunk 1]
```diff
@@ -1,9 +1,15 @@
...
+FLASHFILE ?= $(BINFILE)
+
 RIOT_EDBG = $(RIOTTOOLS)/edbg/edbg
 EDBG ?= $(RIOT_EDBG)
-FLASHER ?= $(EDBG)
-FLASHFILE ?= $(BINFILE)
+ifneq (0,$(shell command -v $(EDBG) 2>&1 > /dev/null ; echo $$?))
+  FLASHDEPS += $(RIOT_EDBG)
+  EDBG := $(RIOT_EDBG)
+endif
```
Comment (fjmolinas) [inline makefiles/tools/edbg.inc.mk:11]: Why did you need to add this? Wouldn't EDBG be part of the dependencies enough?

Comment (aabadie) [reply makefiles/tools/edbg.inc.mk:11]: Good question ! In #15970, for some reason, I always had the message that edbg wasn't found, and the tool Makefile was always triggered. This change fixed it but I agree it's not super nice.

Let me try with the previous logic in this PR.

Comment (aabadie) [reply makefiles/tools/edbg.inc.mk]: Ok so I tried and the problem with the previous logic is that it checks that the FLASHER variable is the same as EDBG. But with PR, FLASHER doesn't correspond anymore to the edbg binary but to the wrapper script.

In the end, it turns out that using the simple:

```mk
EDBG ?= $(RIOTTOOLS)/edbg/edbg
FLASHDEPS += $(EDBG)
```

works the same and is simpler:

- without edbg available in RIOT, it's correctly built
 
```
$ make -C dist/tools/edbg clean  --no-print-directory 
rm -f /work/riot/RIOT/dist/tools/edbg/edbg
$ make BOARD=samr21-xpro -C examples/hello-world flash-only --no-print-directory 
[INFO] edbg binary not found - building it from source now
CC= CFLAGS= make -C /work/riot/RIOT/dist/tools/edbg
env -i PATH="/home/aabadie/.local/bin:/work/tools/mips-mti-elf/2016.05-03/bin:/work/tools/avr8-gnu-toolchain-linux_x86_64/bin:/work/tools/riscv-none-gcc/xpack-riscv-none-embed-gcc-10.1.0-1.1/bin:/work/tools/gcc-arm-none-eabi-9-2019-q4-major/bin/:/home/aabadie/.cargo/bin:/home/aabadie/.cargo/bin:/home/aabadie/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/work/tools/packer" TERM="xterm-256color" "make" -C "/work/riot/RIOT/dist/tools/edbg/bin"
make: Entering directory '/work/riot/RIOT/dist/tools/edbg/bin'
gcc -W -Wall -Wextra -O2 -std=gnu11 dap.c edbg.c target.c target_atmel_cm0p.c target_atmel_cm3.c target_atmel_cm4.c target_atmel_cm7.c target_atmel_cm4v2.c target_mchp_cm23.c target_st_stm32g0.c target_gd_gd32f4xx.c  dbg_lin.c -ludev -o edbg
make: Leaving directory '/work/riot/RIOT/dist/tools/edbg/bin'
mv /work/riot/RIOT/dist/tools/edbg/bin/edbg .
[INFO] edbg binary successfully built!
/work/riot/RIOT/dist/tools/edbg/edbg.sh flash /work/riot/RIOT/examples/hello-world/bin/samr21-xpro/hello-world.bin
### Flashing Target ###
Debugger: ATMEL EDBG CMSIS-DAP ATML2127031800009840 01.1A.00FB (S)
Clock frequency: 16.0 MHz
Target: SAM R21G18 (Rev C)
Verification....................................... done.
Done flashing
```

- if edbg is available in RIOT, it's not re-built
 
```
$ make BOARD=samr21-xpro -C examples/hello-world flash-only --no-print-directory 
/work/riot/RIOT/dist/tools/edbg/edbg.sh flash /work/riot/RIOT/examples/hello-world/bin/samr21-xpro/hello-world.bin
### Flashing Target ###
Debugger: ATMEL EDBG CMSIS-DAP ATML2127031800009840 01.1A.00FB (S)
Clock frequency: 16.0 MHz
Target: SAM R21G18 (Rev C)
Verification....................................... done.
Done flashin
```

File: makefiles/tools/edbg.inc.mk [hunk 2]
```diff
@@ -15,22 +17,21 @@ endif
 
 # Set offset according to IMAGE_OFFSET if it's defined
 EDBG_ARGS += $(if $(IMAGE_OFFSET),--offset $(IMAGE_OFFSET))
+EDBG_ARGS += --target $(EDBG_DEVICE_TYPE)
 
-FFLAGS ?= $(EDBG_ARGS) --target $(EDBG_DEVICE_TYPE) --verbose \
-                       --file $(FLASHFILE)
+EDBG_TARGETS = flash reset
```
Comment (fjmolinas) [inline makefiles/tools/edbg.inc.mk:22]:
```suggestion
EDBG_TARGETS = flash flash-only reset
```

Comment (fjmolinas) [reply makefiles/tools/edbg.inc.mk:22]: ?

**Comment by fjmolinas** [comment]
Tested everything looks good @aabadie please squash!

**Comment by aabadie** [comment]
squashed!

**Review by fjmolinas — APPROVED** [review]
ACK
